File: src/types/emergency.ts
// src/types/emergency.ts - সম্পূর্ণ ঠিক করা version

import { Address } from './user';

// ============================================
// EMERGENCY SERVICE TYPES
// ============================================

export type EmergencyServiceType = 'ambulance' | 'blood' | 'oxygen' | 'doctor' | 'emergency-room' | 'helicopter';

export type EmergencyServiceStatus = 'available' | 'busy' | 'dispatched' | 'offline' | 'maintenance';

export type EmergencyPriority = 'low' | 'medium' | 'high' | 'critical';

export type EmergencyRequestStatus = 'pending' | 'dispatched' | 'en-route' | 'arrived' | 'completed' | 'cancelled';

// ============================================
// EMERGENCY SERVICE INTERFACE
// ============================================
export interface EmergencyService {
  id: string;
  name: string;
  type: EmergencyServiceType;
  status: EmergencyServiceStatus;
  provider: string;
  phone: string;
  location: Address;
  eta?: number; // in minutes
  capacity?: number;
  currentLoad?: number;
  vehicleNumber?: string;
  crewMembers?: number;
  equipment?: string[];
  lastDispatchedAt?: string;
  rating?: number;
  price?: number;
}

// ============================================
// EMERGENCY REQUEST INTERFACE
// ============================================
export interface EmergencyRequest {
  id: string;
  type: EmergencyServiceType;
  priority: EmergencyPriority;
  status: EmergencyRequestStatus;
  patientId?: string;
  patientName?: string;
  patientPhone?: string;
  location: string;
  coordinates?: { lat: number; lng: number };
  notes?: string;
  serviceId?: string;
  serviceName?: string;
  estimatedTime?: number;
  requestedAt: string;
  dispatchedAt?: string;
  arrivedAt?: string;
  completedAt?: string;
  cancelledAt?: string;
  cancellationReason?: string;
}

// ============================================
// EMERGENCY REQUEST PAYLOAD
// ============================================
export interface CreateEmergencyRequestPayload {
  type: EmergencyServiceType;
  priority: EmergencyPriority;
  patientId?: string;
  patientName?: string;
  patientPhone?: string;
  location: string;
  coordinates?: { lat: number; lng: number };
  notes?: string;
}

// ============================================
// EMERGENCY RESOURCE TRACKING
// ============================================
export interface EmergencyResource {
  id: string;
  type: 'ventilator' | 'monitor' | 'defibrillator' | 'stretcher' | 'oxygen-cylinder';
  total: number;
  available: number;
  status: 'operational' | 'low' | 'critical' | 'maintenance';
  lastServiced: string;
}

export interface EmergencyStaff {
  role: 'doctor' | 'nurse' | 'paramedic' | 'driver';
  total: number;
  available: number;
  onCall: number;
  onDuty: number;
}

// ============================================
// EMERGENCY STATE
// ============================================
export interface EmergencyState {
  services: EmergencyService[];
  requests: EmergencyRequest[];
  activeRequests: EmergencyRequest[];
  criticalRequests: EmergencyRequest[];
  resources: EmergencyResource[];
  staff: EmergencyStaff[];
  isLoading: boolean;
  error: string | null;
  isMonitoring: boolean;
}

// ============================================
// EMERGENCY STATS
// ============================================
export interface EmergencyStats {
  totalServices: number;
  availableServices: number;
  activeRequests: number;
  criticalRequests: number;
  avgResponseTime: number; // in minutes
  byType: Record<EmergencyServiceType, number>;
  byPriority: Record<EmergencyPriority, number>;
  successRate: number;
}

File: models/emergency.py
from sqlalchemy import (
    Column, String, Integer, Date, Text, Boolean, DECIMAL, JSON,
    Enum as SQLEnum, ForeignKey, Index
)
from sqlalchemy.dialects.mysql import CHAR
from ..core.database import Base
from .base import BaseModel


# ============================================
# EMERGENCY SERVICES
# ============================================
class EmergencyService(Base, BaseModel):
    __tablename__ = "emergency_services"

    name = Column(String(255), nullable=False)
    type = Column(SQLEnum("ambulance", "blood", "oxygen", "doctor", "emergency-room", "helicopter"), nullable=False, index=True)
    status = Column(SQLEnum("available", "busy", "dispatched", "offline", "maintenance"), default="available", index=True)
    provider = Column(String(255))
    phone = Column(String(20))
    street = Column(String(255))
    city = Column(String(100))
    state = Column(String(100))
    latitude = Column(DECIMAL(10, 7))
    longitude = Column(DECIMAL(10, 7))
    eta_minutes = Column(Integer)
    capacity = Column(Integer)
    current_load = Column(Integer, default=0)
    vehicle_number = Column(String(50))
    crew_members = Column(Integer)
    equipment = Column(JSON)
    rating = Column(DECIMAL(2, 1), default=0.0)
    price = Column(DECIMAL(10, 2), default=0)
    hospital_id = Column(CHAR(36), ForeignKey("hospitals.id"), nullable=True, index=True)
    is_active = Column(Boolean, default=True)


# ============================================
# EMERGENCY REQUESTS
# ============================================
class EmergencyRequest(Base, BaseModel):
    __tablename__ = "emergency_requests"

    type = Column(SQLEnum("ambulance", "blood", "oxygen", "doctor", "emergency-room"), nullable=False, index=True)
    priority = Column(SQLEnum("low", "medium", "high", "critical"), default="medium", index=True)
    status = Column(SQLEnum("pending", "dispatched", "en-route", "arrived", "completed", "cancelled"), default="pending", index=True)
    patient_id = Column(CHAR(36), ForeignKey("users.id"), nullable=True, index=True)
    patient_name = Column(String(150))
    patient_phone = Column(String(20))
    location_address = Column(Text, nullable=False)
    latitude = Column(DECIMAL(10, 7))
    longitude = Column(DECIMAL(10, 7))
    notes = Column(Text)
    service_id = Column(CHAR(36), ForeignKey("emergency_services.id"), nullable=True)
    hospital_id = Column(CHAR(36), ForeignKey("hospitals.id"), nullable=True)
    estimated_time = Column(Integer, comment="ETA in minutes")
    dispatched_at = Column(Date, nullable=True)
    arrived_at = Column(Date, nullable=True)
    completed_at = Column(Date, nullable=True)
    cancelled_at = Column(Date, nullable=True)
    cancellation_reason = Column(String(500))
